src/data/pi/lex.lean

```diff
@@ -63,36 +80,29 @@ instance lex.is_strict_order [linear_order ι] [∀ a, partial_order (β a)] :
 instance [linear_order ι] [Π a, partial_order (β a)] : partial_order (lex (Π i, β i)) :=
 partial_order_of_SO (<)
 
-protected lemma lex.is_strict_total_order' [linear_order ι]
-  (wf : well_founded ((<) : ι → ι → Prop)) [∀ a, linear_order (β a)] :
-  is_strict_total_order' (lex (Π i, β i)) (<) :=
-{ trichotomous := λ a b,
-    begin
-      by_cases h : ∃ i, a i ≠ b i,
-      { let i := wf.min _ h,
-        have hlt_i : ∀ j < i, a j = b j,
-        { intro j, rw ← not_imp_not,
-          exact λ h', wf.not_lt_min _ _ h' },
-        have : a i ≠ b i, from wf.min_mem _ h,
-        exact this.lt_or_lt.imp (λ h, ⟨i, hlt_i, h⟩)
-          (λ h, or.inr ⟨i, λ j hj, (hlt_i j hj).symm, h⟩) },
-      { push_neg at h, exact or.inr (or.inl (funext h)) }
-    end }
-
-/-- `Πₗ i, α i` is a linear order if the original order is well-founded.
-This cannot be an instance, since it depends on the well-foundedness of `<`. -/
-protected noncomputable def lex.linear_order [linear_order ι]
-  (wf : well_founded ((<) : ι → ι → Prop)) [∀ a, linear_order (β a)] :
+/-- `Πₗ i, α i` is a linear order if the original order is well-founded. -/
+noncomputable instance [linear_order ι] [is_well_order ι (<)] [∀ a, linear_order (β a)] :
   linear_order (lex (Π i, β i)) :=
-@linear_order_of_STO' (Πₗ i, β i) (<) (pi.lex.is_strict_total_order' wf) (classical.dec_rel _)
-
-lemma lex.le_of_forall_le [linear_order ι]
-  (wf : well_founded ((<) : ι → ι → Prop)) [Π a, linear_order (β a)] {a b : lex (Π i, β i)}
-  (h : ∀ i, a i ≤ b i) : a ≤ b :=
-begin
-  letI : linear_order (Πₗ i, β i) := lex.linear_order wf,
-  exact le_of_not_lt (λ ⟨i, hi⟩, (h i).not_lt hi.2)
-end
+@linear_order_of_STO' (Πₗ i, β i) (<)
+  { to_is_trichotomous := is_trichotomous_lex _ _ is_well_order.wf } (classical.dec_rel _)
+
+lemma lex.le_of_forall_le [linear_order ι] [is_well_order ι (<)] [Π a, linear_order (β a)]
```

Comment:
Can you prove `monotone to_lex` instead? This will shorten the `order_bot` and `order_top` instances (although I'm surprised we only have the pullbacks `order_bot.lift`/`order_top.lift` and pushforward).

Reply:
Added.

Reply:
I think you can even get rid of `lex.le_of_forall_le` because it's just an inconvenient spelling of `to_lex_monotone` now.

Reply:
What we could do is change the statement here from`(h : ∀ i, a i ≤ b i) : a ≤ b` to `(h : of_lex a ≤ of_lex b) : a ≤ b`

Reply:
Since we accept `(f : Πₗ i, β i) i` anyway, let's have all these versions.